homeassistant/components/esphomelib/__init__.py

```diff
@@ -0,0 +1,372 @@
+    entry_data.info[component_key] = {}
+    entry_data.state[component_key] = {}
+
+    @callback
+    def async_list_entities(infos: List[Any]):
+        """Update entities of this platform when entities are listed."""
+        old_infos = entry_data.info[component_key]
+        new_infos = {}
+        add_entities = []
+        for info in infos:
+            if not isinstance(info, info_type):
+                # Filter out infos that don't belong to this platform.
+                continue
+
+            if info.key in old_infos:
+                # Update existing entity
+                old_infos.pop(info.key)
+            else:
+                # Create new entity
+                entity = entity_type(entry.entry_id, component_key, info.key)
+                add_entities.append(entity)
+            new_infos[info.key] = info
+
+        # Remove old entities
+        for info in old_infos.values():
+            entry_data.async_remove_entity(hass, component_key, info.key)
+        entry_data.info[component_key] = new_infos
+        async_add_entities(add_entities)
+
+    signal = DISPATCHER_ON_LIST.format(entry_id=entry.entry_id)
+    async_dispatcher_connect(hass, signal, async_list_entities)
+
+    @callback
+    def async_entity_state(state):
+        """Notify the appropriate entity of an updated state."""
+        if not isinstance(state, state_type):
+            return
+        entry_data.state[component_key][state.key] = state
+        async_dispatcher_send(
+            hass, DISPATCHER_UPDATE_ENTITY.format(entry_id=entry.entry_id,
+                                                  component_key=component_key,
+                                                  key=state.key))
+
+    signal = DISPATCHER_ON_STATE.format(entry_id=entry.entry_id)
+    async_dispatcher_connect(hass, signal, async_entity_state)
+
+
+class EsphomelibEntity(Entity):
+    """Define a generic esphomelib entity."""
+
+    def __init__(self, entry_id, component_key, key):
+        """Initialize."""
+        self._entry_id = entry_id
+        self._component_key = component_key
+        self._key = key
+
+    async def async_added_to_hass(self):
+        """Register callbacks."""
+        kwargs = {
+            'entry_id': self._entry_id,
+            'component_key': self._component_key,
+            'key': self._key,
+        }
+        async_dispatcher_connect(self.hass,
+                                 DISPATCHER_UPDATE_ENTITY.format(**kwargs),
+                                 self.async_schedule_update_ha_state)
+
+        async_dispatcher_connect(self.hass,
+                                 DISPATCHER_REMOVE_ENTITY.format(**kwargs),
+                                 self.async_remove)
+
+        async_dispatcher_connect(self.hass,
+                                 DISPATCHER_ON_DEVICE_UPDATE.format(**kwargs),
+                                 self.async_schedule_update_ha_state)
+
+    @property
```

Comment:
Add a coroutine method `async_will_remove_from_hass` here that disconnects the dispatchers. When connecting to the dispatcher signal a disconnect function is returned. Save those eg in a list as an instance attribute.

Reply:
See rainmachine component for example.